hail/python/hail/expr/expressions/typed_expressions.py

```diff
@@ -3347,6 +3347,59 @@ def __floordiv__(self, other):
     def __rfloordiv__(self, other):
         return self._bin_op_numeric_reverse('//', other)
 
+    def __matmul__(self, other):
+        """Matrix multiplication: `a @ b`, semantically equivalent to `NumPy` matmul. If `a` and `b` are vectors,
+        the vector dot product is performed, returning a `NumericExpression`. If `a` and `b` are both 2-dimensional
+        matrices, this performs normal matrix multiplication. If `a` and `b` have more than 2 dimensions, they are
+        treated as multi-dimensional stacks of 2-dimensional matrices. Matrix multiplication is applied element-wise
+        across the higher dimensions. E.g. if `a` has shape `(3, 4, 5)` and `b` has shape `(3, 5, 6)`, `a` is treated
+        as a stack of three matrices of shape `(4, 5)` and `b` as a stack of three matrices of shape `(5, 6)`. `a @ b`
+        would then have shape `(3, 4, 6)`.
+
+        Notes
+        -----
+        The last dimension of `a` and the second to last dimension of `b` (or only dimension if `b` is a vector)
+        must have the same length. The dimensions to the left of the last two dimensions of `a` and `b` (for NDArrays
+        of dimensionality > 2) must be equal or be compatible for broadcasting.
+        Number of dimensions of both NDArrays must be at least 1.
+
+        Parameters
+        ----------
+        other : :class:`numpy.ndarray` :class:`.NDArrayNumericExpression`
+
+        Returns
+        -------
+        :class:`.NDArrayNumericExpression` or :class:`.NumericExpression`
+        """
+        if not isinstance(other, NDArrayNumericExpression):
+            other = hl._ndarray(other)
```

Comment:
@tpoterba can comment, but we should hook into the @typecheckmethod functionality to get this automatically and across any method that takes ndarray.

Reply:
yes, I agree.

Reply:
That would be ideal, and as it stands right now there's a bit of a snag. Inputs to type checked methods are always first coerced with `to_expr` with the assumption that there's no ambiguity between python types and hail types. So we can easily coerce numpy ndarrays but nested python lists would be coerced to hail arrays of hail arrays. That might be fine, and I can implement that, but you would no longer be able to interop implicitly with python lists. Could also change how type coercion works but that would be less simple.

Reply:
Hmm. @tpoterba do you want to talk with @daniel-goldstein on Tuesday about how we can expand to_expr to handle this situation?

Reply:
I should we should be able to write an ndarray coercer that can coerce python lists

Reply:
Sounds good!